### PR TITLE
Resolve ILLink warnings in System.ComponentModel.TypeConverter (Round 1)

### PR DESCRIPTION
**Pull request by eerhardt**
Contributes to #45623

This change got a little big, so I figured I'd break it off into its own PR for easier reviewing.

### CODE REVIEW

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs [hunk 1]
```diff
@@ -212,15 +212,19 @@ protected override void FillAttributes(IList attributeList)
             base.FillAttributes(attributeList);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "No Attributes are supplied.")]
```
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/PropertyDescriptor.cs:215]: @vitek-karas @MichalStrehovsky - this is a scenario where it would be really helpful if the library could define its own warning number for `RequiresUnreferencedCode`.

Today, passing an `Attribute[] filter` is what requires unreferenced code. But in the future, if `GetChildProperties` gets marked unsafe for a new reason, this suppression will still apply, and we won't get a warning.

So if I could mark `public virtual PropertyDescriptorCollection GetChildProperties(object instance, Attribute[] filter)` as `[RequiresUnreferencedCode(AttributeCollection.FilterRequiresUnreferencedCodeMessage, "SYSLIB12345")]`, then I could suppress just `SYSLIB12345` here. And if another `RequiresUnreferencedCode` gets added in the call stack in the future, this would start warning for the new warning.

Thoughts?

**Comment by eerhardt** [comment]
Hold off on reviewing this. Changes from the next round are going to invalidate what I've done here. So this change is going to have to get bigger.

**Comment by eerhardt** [comment]
This is now ready to be reviewed if you want to take a look

@joperezr @tannergooding @MichalStrehovsky @vitek-karas 

I still need 1 more change to update the ref assemblies for `DynamicallyAccessedMembers`. There is currently a bug in API compat I need fixed for it.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs [hunk 2]
```diff
@@ -29,6 +33,7 @@ public AssociatedMetadataTypeTypeDescriptor(ICustomTypeDescriptor parent, Type t
             }
         }
 
+        [RequiresUnreferencedCode("The public parameterless constructor or the 'Default' static field may be trimmed from the Attribute's Type.")]
```
Comment (joperezr) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs:36]: Don't we also need an entry on the ILLink.LibraryBuild.xml for every RequiresUnreferencedCode that we add?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ComponentResourceManager.cs [hunk 3]
```diff
@@ -51,6 +52,7 @@ private CultureInfo NeutralResourcesCulture
         /// to the corresponding property on the object. If there is no matching
         /// property the resource will be ignored.
         /// </summary>
+        [RequiresUnreferencedCode("The Type of value cannot be statically discovered.")]
```
Comment (joperezr) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ComponentResourceManager.cs:55]: NIT: Share the string of both of these API's attributes?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectPropertyDescriptor.cs [hunk 4]
```diff
@@ -221,7 +246,7 @@ private EventDescriptor ChangedEventValue
             {
                 if (!_state[s_bitChangedQueried])
                 {
-                    _realChangedEvent = TypeDescriptor.GetEvents(ComponentType)[Name + "Changed"];
+                    _realChangedEvent = TypeDescriptor.GetEvents(_componentClass)[Name + "Changed"];
```
Comment (joperezr) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectPropertyDescriptor.cs:249]: Shouldn't we also annotate the PropertyGetter?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs [hunk 5]
```diff
@@ -172,11 +173,19 @@ public TypeConverter GetOrCreateConverterInstance(Type innerType)
...
             [typeof(Array)] = new IntrinsicTypeConverterData((type) => new ArrayConverter()),
             [typeof(ICollection)] = new IntrinsicTypeConverterData((type) => new CollectionConverter()),
-            [typeof(Enum)] = new IntrinsicTypeConverterData((type) => new EnumConverter(type), cacheConverterInstance: false),
+            [typeof(Enum)] = new IntrinsicTypeConverterData((type) => CreateEnumConverter(type), cacheConverterInstance: false),
             [s_intrinsicNullableKey] = new IntrinsicTypeConverterData((type) => new NullableConverter(type), cacheConverterInstance: false),
             [s_intrinsicReferenceKey] = new IntrinsicTypeConverterData((type) => new ReferenceConverter(type), cacheConverterInstance: false),
         });
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2067:UnrecognizedReflectionPattern",
+            Justification = "Trimmer does not trim enums")]
```
Comment (joperezr) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:182]: Is it possible for this to change in future versions of the trimmer? If so what is the plan?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs [hunk 6]
```diff
@@ -232,6 +233,8 @@ protected Exception GetConvertToException(object value, Type destinationType)
         /// the specified context.
         ///
         /// </summary>
+        [RequiresUnreferencedCode("The Type of value cannot be statically discovered.")]
+        [DynamicDependency(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor | DynamicallyAccessedMemberTypes.PublicFields, typeof(BrowsableAttribute))]
```
Comment (joperezr) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:237]: just for clarification: is the idea of having a dynamicDependency attribute here just to do best effort of preserving what is needed? I ask since it also has RequiresUnreferencedCode so we know it won't be safe anyway.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs [hunk 7]
```diff
@@ -364,7 +368,13 @@ protected SimplePropertyDescriptor(Type componentType, string name, Type propert
             /// <summary>
             /// Gets a value indicating whether this property is read-only.
             /// </summary>
-            public override bool IsReadOnly => Attributes.Contains(ReadOnlyAttribute.Yes);
+
+            public override bool IsReadOnly
+            {
+                [DynamicDependency(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor | DynamicallyAccessedMemberTypes.PublicFields, typeof(ReadOnlyAttribute))]
+                [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "The DynamicDependency ensures the correct members are preserved.")]
```
Comment (joperezr) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:375]: Specifically for cases like this and other places where you are suppressing warnings, shouldn't we have a test for that to ensure it is true? I ask since I don't really see Trimming tests added by this PR.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs (hunk 5 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs:182]: I don't believe it is possible for the trimmer to change this behavior. We make the same justification elsewhere:

https://github.com/dotnet/runtime/blob/ecc354a61c909a1960af7e5a73eb9a0d255c564e/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L5510-L5519

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs:36]: No, we only need those entries if there are ILLink warnings that aren't being addressed. In this case, we are marking all calls to this API as `RequiresUnreferencedCode` as well, so the trimmer doesn't warn.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectPropertyDescriptor.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectPropertyDescriptor.cs:249]: We can't because it is an override. If you apply it to a virtual property, it needs to be applied to the whole hierarchy.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ComponentResourceManager.cs (hunk 3 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ComponentResourceManager.cs:55]: In some cases it helps sharing strings (like the AttributeCollection case). For these strings, I found that sometimes the method uses the word `value`, sometimes `instance`, and sometimes, `component`. So I just started copying the strings because I didn't want to reference the wrong word.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs (hunk 6 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:237]: This method is doing 2 things that are "unsafe".

1. It is filtering attributes
2. It is calling `GetProperties` passing in an object, which eventually gets `GetType()` called on it.

We can make the first thing "safe" by applying the DynamicDependency. So calling this method will preserve the correct things on the `BrowsableAttribute` class. But the 2nd thing can't be made safe, so it is marked with `RUC`. Consumers of this method who wish to make it safe, can manually mark `value`'s Type as needed, but they would need to dig into the code to figure out why `BrowsableAttribute` is broken if we didn't use `DynamicDependency`.

So in short, this makes it as "safe" as we can make it.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs (hunk 7 above)
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:375]: A lot of the value we got from writing trimming tests previously was because we didn't have the linker analysis to tell us what is and isn't safe. But we do now, so we are relying heavily on the linker.

I can add a test for this.

**Review by vitek-karas — APPROVED** [review]
LGTM - ideally somebody else should also approve - it's a big change :-)

**Review by joperezr — APPROVED** [review]
